xhttp/client.go

```diff
@@ -99,7 +109,9 @@ func (r *retrierClient) do(ctx context.Context, req *http.Request, requestBody [
 
 	log := slog.FromCtx(ctx).With("request_url", req.URL)
 
+	start := time.Now()
 	res, err := r.client.Do(req)
+	r.onRequestDone(req, res, err, time.Since(start))
```

Comment:
where's the guarantee for the invariant that `res == nil` imply `err != nil` and vice versa? unless the `r.client` is not Go's HTTP client, in that case, ignore.

Reply:
hmm I assumed this is guaranteed by the underlying http.Do that we wrap, usually when you have an error you should ignore the returned value..but indeed not always. I was re-reading the [docs](https://pkg.go.dev/net/http#Client.Do)...and actually they don't explicitly say that the res will be nil, there is only an invariant related to the error being nil and the body being always non-nil. Maybe it is better to just document that we will pass on the callback what is returned by the Do method instead of trying to duplicate/guarantee invariants, this is supposed to be very lightweight and just give you a view on what was returned by Do/underlying request.

Reply:
> for context: https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/net/http/client.go;l=710-719

Just now saw the link =P. Yeah I brainfarted because of the invariant of "when the error is nil the body is always non-nil", I think I extrapolated some other stuff from that.

In the end I'm not sure we should enforce a lot of invariants, it all depends on the underlying http.Client...usually it is the Go one...but it could be a different one...we don't enforce that.

Reply:
> In the end I'm not sure we should enforce a lot of invariants, it all depends on the underlying http.Client...usually it is the Go one...but it could be a different one...we don't enforce that.

makes sense.

Reply:
Yeah the docs also say that actually, so just read the docs I suppose :rofl: 

"A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned [Response.Body] is already closed."

Reply:
It is quite complex...in the end..this whole protocol (at least to me..given my obvious limitations :rofl: )...because every time you have a valid response you are supposed to close the body..and body won't be nil, and usually when there is an error there is no response...but when there is an error of a specific kind then there will be a response..but in that case you don't need to close the body :laughing: (and it is safe to ignore it...as it is usual on Go's error handling).

Reply:
yep... as usual, because they need to maintain compatibility with Go 1, they make things more complicated than needed. For almost everything else we take this for granted, that if `err != nil` then actual value/response is nil/empty.
The cognitive load of maintaining software just increases...
I see the same problem with the `io.Reader` interface ... some many cases to handle because of past bad decisions...

Reply:
no problem with bad decisions... they happen but improving is important.

Reply:
io.Reader and now the http.Client.Do method always make me wonder how code is not enough to specify these protocols, you need out of band natural language specifications to document the protocol (and it is harder to detect bugs).